README.md
# Next Transition Bar

Elevate the user experience in your Next.js applications effortlessly with **next-transition-bar** – a versatile npm package designed to seamlessly enhance page transition progress. This lightweight and customizable solution adds a top loader and progress bar, delivering both visual appeal and a smooth transition experience.

[Live Demo](https://next-transition-bar.vercel.app)

## Installation

You can install the package using npm:

```bash
npm install next-transition-bar
```

Or if you prefer using yarn:

```bash
yarn add next-transition-bar
```

## Usage

Start by importing the package:

```js
import NextTransitionBar from 'next-transition-bar';
```

### Integration with `app/layout.js` (for `app` folder structure)

Include `<NextTransitionBar />` within the `return()` statement inside the `<body></body>` of your `RootLayout()`:

```js
import NextTransitionBar from 'next-transition-bar';

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>
        <NextTransitionBar />
        {children}
      </body>
    </html>
  );
}
```

### Integration with `pages/_app.js` (for `pages` folder structure)

To render, add `<NextTransitionBar />` within the `return()` statement in `MyApp()`:

```js
import NextTransitionBar from 'next-transition-bar';

export default function MyApp({ Component, pageProps }) {
  return (
    <>
      <NextTransitionBar />
      <Component {...pageProps} />;
    </>
  );
}
```

### Importing nprogress instance

To import `nprogress` from 'next-transition-bar':

```js
import { nprogress } from 'next-transition-bar';
```

or both

```js
import NextTransitionBar, { nprogress } from 'next-transition-bar';
```

### Default Configuration

If no props are passed to `<NextTransitionBar />`, the package applies the following default configuration:

```jsx
<NextTransitionBar
  color="#29d"
  initialPosition={0.08}
  trickleSpeed={200}
  height={3}
  trickle={true}
  showSpinner={true}
  easing="ease"
  speed={200}
  shadow="0 0 10px #29d, 0 0 5px #29d"
  template='<div class="bar" role="bar"><div class="peg"></div></div>
            <div class="spinner" role="spinner"><div class="spinner-icon"></div></div>'
  zIndex={1600}
  showAtBottom={false}
  isRTL={false}
  nonce={undefined}
  transformCSS={(css) => <style nonce={undefined}>{css}</style>}
/>
```

- `color`: Change the default color of the top loader.
- `initialPosition`: Adjust the initial position of the top loader (in percentage, e.g., `0.08 = 8%`).
- `trickleSpeed`: Incremental delay speed in milliseconds.
- `speed`: Animation speed for the top loader in milliseconds.
- `easing`: Animation settings using easing (a CSS easing string).
- `height`: Height of the top loader in pixels.
- `trickle`: Auto-incrementing behavior for the top loader.
- `showSpinner`: Toggle spinner visibility.
- `shadow`: A smooth shadow for the top loader (set to `false` to disable it).
- `template`: Include custom HTML attributes for the top loader.
- `zIndex`: Define zIndex for the top loader.
- `showAtBottom`: Display the top loader at the bottom (increase height to ensure visibility on mobile devices).
- `isRTL`: Change the direction of the loader.
- `nonce`: The nonce attribute to use for the `style` tag.
- `transformCSS`: This is useful if you want to use a different style or minify the CSS.

Experience a sleek and visually appealing progress bar with **next-transition-bar**. Customize it to suit your application's unique style and provide users with a delightful journey through your Next.js app.


CHANGELOG.md
# [1.1.0](https://github.com/devmahmud/next-transition-bar/compare/v1.0.1...v1.1.0) (2023-11-21)


### Bug Fixes

* name ([0644a1b](https://github.com/devmahmud/next-transition-bar/commit/0644a1be87349c662b161fc3369459ae285dc856))


### Features

* added tests ([0bcd9b8](https://github.com/devmahmud/next-transition-bar/commit/0bcd9b8b40af2ece9f4bfce8f2fce6351df1403a))
* added tests ([3ca5831](https://github.com/devmahmud/next-transition-bar/commit/3ca58318e5d83d1b55bdac34801f06797239d660))


tests/nprogress.test.ts
import { describe, test, expect, vi } from 'vitest';
import nprogress from '../src/nprogress';
import { afterEach } from 'node:test';

const defaults = {
  minimum: 0.08,
  easing: 'linear',
  speed: 200,
  trickle: true,
  trickleSpeed: 200,
  showSpinner: true,
  isRTL: false,
  barSelector: 'div.bar',
  barLabel: 'processing request',
  spinnerSelector: 'div.spinner',
  spinnerLabel: 'processing request',
  parent: 'body',
  template: `
    <div class="bar" role="progressbar" aria-valuemin="0" aria-valuemax="1">
      <div class="peg"></div>
    </div>
    <div class="spinner" role="progressbar" aria-valuemin="0" aria-valuemax="1">
      <div class="spinner-icon"></div>
    </div>
  `,
};

vi.useFakeTimers();

describe('NProgress', () => {
  afterEach(() => {
    nprogress.done();
    const element = document.getElementById('nprogress');
    if (element) {
      element.remove();
    }
    nprogress.configure(defaults);
  });

  test('should have defaults', () => {
    expect(nprogress.status).toBe(null);
    expect(nprogress.settings).toEqual(defaults);
  });

  describe('set', () => {
    test('should render nprogress', () => {
      nprogress.set(0);
      expect(document.getElementById('nprogress')).toBeDefined();
      expect(document.querySelector('#nprogress .bar')).toBeDefined();
      expect(document.querySelector('#nprogress .peg')).toBeDefined();
      expect(document.querySelector('#nprogress .spinner')).toBeDefined();
    });

    test('should appear and disappear', () => {
      nprogress.configure({ speed: 10 });
      nprogress.set(0).set(1);
      vi.runAllTimers();
      expect(document.getElementById('nprogress')).toBeDefined();
      expect(document.getElementById('nprogress')).toEqual(null);
    });

    test('should respect minimum', () => {
      nprogress.set(0);
      expect(nprogress.status).toEqual(defaults.minimum);
    });

    test('should clamp to minimum', () => {
      nprogress.set(-100);
      expect(nprogress.status).toEqual(defaults.minimum);
    });

    test('must clamp to maximum', () => {
      nprogress.set(456);
      expect(nprogress.status).toEqual(null);
    });
  });

  describe('start', () => {
    test('should render', () => {
      nprogress.start();
      expect(document.getElementById('nprogress')).toBeDefined();
    });

    test('should respect minimum', () => {
      nprogress.start();
      expect(nprogress.status).toEqual(defaults.minimum);
    });

    test('should be attached to specified parent when parent is already in the DOM', () => {
      document.body.innerHTML = '<div id="test"></div>';
      nprogress.configure({ parent: '#test' });
      nprogress.start();
      // expect(document.getElementById('nprogress')?.parentElement).toEqual(document.getElementById('test'));
      // expect(document.getElementById('test')?.classList[0]).toEqual('nprogress-custom-parent');
    });

    test('should be attached to specified parent when parent is delayed being rendered to the DOM', () => {
      nprogress.configure({ parent: '#test' });
      nprogress.start();
      document.body.innerHTML = '<div id="test"></div>';

      expect(document.getElementById('nprogress')).toBeNull();

      vi.advanceTimersByTime(1000);

      expect(document.getElementById('nprogress')?.parentElement).toEqual(document.getElementById('test'));
      expect(document.getElementById('test')?.classList[0]).toEqual('nprogress-custom-parent');
    });

    test('should increment status until done', () => {
      nprogress.start();
      vi.runOnlyPendingTimers();
      expect(nprogress.status).toBeGreaterThan(defaults.minimum);
      nprogress.done();
      vi.runAllTimers();
      expect(nprogress.status).toBeNull();
    });
  });

  describe('done', () => {
    test('should not render without start', () => {
      nprogress.done();
      expect(document.getElementById('nprogress')).toBeNull();
    });

    test('done(true) should render', () => {
      nprogress.done(true);
      expect(document.getElementById('nprogress')).toBeDefined();
    });
  });

  describe('remove', () => {
    test('should be removed from the parent', () => {
      document.body.innerHTML = '<div id="test"></div>';
      nprogress.configure({ parent: '#test' });
      nprogress.set(1);
      expect(document.getElementById('nprogress')).toBeDefined();
      expect(document.getElementById('test')?.classList[0]).toEqual('nprogress-custom-parent');

      nprogress.remove();

      expect(document.getElementById('nprogress')).toBeNull();
      expect(document.getElementById('test')?.classList.length).toEqual(0);
    });
  });

  describe('inc', () => {
    test('should render', () => {
      nprogress.inc();
      expect(document.getElementById('nprogress')).toBeDefined();
    });

    test('should start with minimum', () => {
      expect(nprogress.status).toEqual(nprogress.settings.minimum);
    });

    test('should increment', () => {
      nprogress.start();
      const start = nprogress.status;

      nprogress.inc();
      expect(nprogress.status).toBeGreaterThan(Number(start));
    });

    test('should never reach 1.0', () => {
      for (let i = 0; i < 100; i += 1) {
        nprogress.inc();
      }
      expect(nprogress.status).toBeLessThan(1.0);
    });
  });

  describe('configure', () => {
    test('should override defaults', () => {
      nprogress.configure({ minimum: 0.5 });
      expect(nprogress.settings.minimum).toEqual(0.5);
    });
  });

  describe('configure(showSpinner)', () => {
    test('should render spinner by default', () => {
      nprogress.start();
      expect(document.querySelector('#nprogress .spinner')).toBeDefined();
    });

    test('should hide (on false)', () => {
      document.getElementById('nprogress')?.remove();
      nprogress.configure({ showSpinner: false });
      nprogress.start();
      expect(document.querySelector('#nprogress')).toBeDefined();
      expect(document.querySelector('#nprogress .spinner')).toBeNull();
    });
  });

  describe('promise', () => {
    test('should set status for promises added', () => {
      nprogress.promise(Promise.resolve());
      nprogress.promise(Promise.resolve());

      expect(document.getElementById('nprogress')).toBeDefined();
    });
  });
});


src/nprogress.ts
interface NProgressSettings {
  minimum: number;
  easing: string;
  speed: number;
  trickle: boolean;
  trickleSpeed: number;
  showSpinner: boolean;
  barSelector: string;
  barLabel: string;
  spinnerSelector: string;
  spinnerLabel: string;
  parent: string;
  template: string;
  isRTL: boolean;
}

interface NProgressInstance {
  configure(options: Partial<NProgressSettings>): NProgressInstance;
  set(value: number): NProgressInstance;
  start(): NProgressInstance;
  done(force?: boolean): NProgressInstance;
  inc(amount?: number): NProgressInstance;
  remove(): void;
  promise($promise: Promise<unknown>): NProgressInstance;
  readonly status: number | null;
  readonly settings: NProgressSettings;
}

const DEFAULTS: NProgressSettings = {
  minimum: 0.08,
  easing: 'linear',
  speed: 200,
  trickle: true,
  trickleSpeed: 200,
  showSpinner: true,
  isRTL: false,
  barSelector: 'div.bar',
  barLabel: 'processing request',
  spinnerSelector: 'div.spinner',
  spinnerLabel: 'processing request',
  parent: 'body',
  template: `
    <div class="bar" role="progressbar" aria-valuemin="0" aria-valuemax="1">
      <div class="peg"></div>
    </div>
    <div class="spinner" role="progressbar" aria-valuemin="0" aria-valuemax="1">
      <div class="spinner-icon"></div>
    </div>
  `,
};

const NProgress = (): NProgressInstance => {
  const localSettings: NProgressSettings = DEFAULTS;
  let localStatus: number | null = null;
  let initialPromises = 0;
  let currentPromises = 0;

  function isRendered(): boolean {
    return !!document.getElementById('nprogress');
  }

  function isStarted(): boolean {
    return typeof localStatus === 'number';
  }

  function render(): HTMLElement {
    if (isRendered()) {
      return document.getElementById('nprogress') as HTMLElement;
    }

    document.documentElement.classList.add('nprogress-busy');

    const progress = document.createElement('div');
    progress.id = 'nprogress';
    progress.innerHTML = localSettings.template;

    const percentage = isStarted() ? `${localSettings.isRTL ? '100' : '-100'}` : toBarPercentage(localStatus || 0);
    const bar = progress.querySelector(localSettings.barSelector) as HTMLElement;
    bar.setAttribute('aria-label', localSettings.barLabel);
    bar.style.transform = `translate3d(${percentage}%,0,0)`;
    bar.style.transition = 'all 0 linear';

    const spinner = progress.querySelector(localSettings.spinnerSelector) as HTMLElement;
    if (spinner) {
      if (!localSettings.showSpinner) {
        removeElement(spinner);
      } else {
        spinner.setAttribute('aria-label', localSettings.spinnerLabel);
      }
    }

    const parent = document.querySelector(localSettings.parent);
    if (parent) {
      if (parent !== document.body) {
        parent.classList.add('nprogress-custom-parent');
      }

      parent.appendChild(progress);
    }
    return progress;
  }

  function clamp(value: number, min: number, max: number): number {
    if (value < min) return min;
    if (value > max) return max;
    return value;
  }

  function toBarPercentage(value: number): number {
    return localSettings.isRTL ? 100 - value * 100 : (-1 + value) * 100;
  }

  function randomInc(status: number | null): number {
    if (typeof status !== 'number') {
      return 0;
    }
    if (status >= 0 && status < 0.2) {
      return 0.1;
    }
    if (status >= 0.2 && status < 0.5) {
      return 0.04;
    }
    if (status >= 0.5 && status < 0.8) {
      return 0.02;
    }
    if (status >= 0.8 && status < 0.99) {
      return 0.005;
    }
    return 0;
  }

  function removeElement(element: HTMLElement): void {
    if (element && element.parentNode) {
      element.parentNode.removeChild(element);
    }
  }

  const queue = (() => {
    const functionQueue: (() => void)[] = [];

    function next(): void {
      const fn = functionQueue.shift();
      if (fn) {
        fn();
      }
    }

    return (fn: (next: () => void) => void): void => {
      functionQueue.push(() => {
        fn(next);
      });

      if (functionQueue.length === 1) {
        next();
      }
    };
  })();

  return {
    configure(options: Partial<NProgressSettings>): NProgressInstance {
      for (const _key in options) {
        const key = _key as keyof NProgressSettings;
        if (Object.prototype.hasOwnProperty.call(options, key) && Object.prototype.hasOwnProperty.call(DEFAULTS, key)) {
          const value = options[key];
          if (value !== undefined) {
            localSettings[key] = value as never;
          }
        }
      }
      return this;
    },

    set(value: number): NProgressInstance {
      const clamppedValue = clamp(value, localSettings.minimum, 1);
      localStatus = clamppedValue === 1 ? null : clamppedValue;

      const progress = render();

      progress.offsetWidth;

      queue((next) => {
        const { speed, easing } = localSettings;
        const bar = progress.querySelector(localSettings.barSelector) as HTMLElement;
        bar.setAttribute('aria-valuenow', clamppedValue.toString());
        bar.style.transform = `translate3d(${toBarPercentage(clamppedValue)}%,0,0)`;
        bar.style.transition = `all ${speed}ms ${easing}`;

        if (clamppedValue === 1) {
          progress.style.transition = 'none';
          progress.style.opacity = '1';

          progress.offsetWidth;

          setTimeout(() => {
            progress.style.transition = `all ${speed}ms linear`;
            progress.style.opacity = '0';
            setTimeout(() => {
              this.remove();
              next();
            }, speed);
          }, speed);
        } else {
          setTimeout(next, speed);
        }
      });

      return this;
    },

    start(): NProgressInstance {
      if (!localStatus) {
        this.set(0);
      }

      const work = () => {
        setTimeout(() => {
          if (!localStatus) {
            return;
          }
          this.inc();
          work();
        }, localSettings.trickleSpeed);
      };

      if (localSettings.trickle) {
        work();
      }

      return this;
    },

    done(force?: boolean): NProgressInstance {
      if (!force && !localStatus) {
        return this;
      }

      const halfRandom = 0.5 * Math.random();
      return this.inc(0.3 + halfRandom).set(1);
    },

    inc(amount: number = randomInc(localStatus)): NProgressInstance {
      if (!localStatus) {
        return this.start();
      }

      const clamppedStatus = clamp(localStatus + amount, 0, 0.994);
      return this.set(clamppedStatus);
    },

    remove(): void {
      document.documentElement.classList.remove('nprogress-busy');
      (document.querySelector(localSettings.parent) as HTMLElement).classList.remove('nprogress-custom-parent');
      const progress = document.getElementById('nprogress');
      if (progress) {
        removeElement(progress);
      }
    },

    promise($promise: Promise<unknown>): NProgressInstance {
      if (currentPromises === 0) {
        this.start();
      }

      initialPromises += 1;
      currentPromises += 1;

      const promiseResolution = () => {
        currentPromises -= 1;
        if (currentPromises === 0) {
          initialPromises = 0;
          this.done();
        } else {
          this.set((initialPromises - currentPromises) / initialPromises);
        }
      };

      $promise.then(promiseResolution).catch(promiseResolution);

      return this;
    },

    get status(): number | null {
      return localStatus;
    },

    get settings(): NProgressSettings {
      return localSettings;
    },
  };
};

export default NProgress();
